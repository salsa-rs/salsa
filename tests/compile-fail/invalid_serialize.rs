#[salsa::tracked(serialize)]
struct Serializable<'db> {
    field: NotSerializable<'db>,
}

#[salsa::tracked]
struct NotSerializable<'db> {
    field: usize,
}

#[salsa::tracked(serialize)]
fn query(_db: &dyn salsa::Database, _input: NotSerializable<'_>) {}

fn main() {}
