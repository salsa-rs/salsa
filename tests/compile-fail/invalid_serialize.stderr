error[E0277]: the trait bound `NotSerializable<'_>: Serialize` is not satisfied
 --> tests/compile-fail/invalid_serialize.rs:1:1
  |
1 | #[salsa::tracked(serialize)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  | |
  | the trait `Serialize` is not implemented for `NotSerializable<'_>`
  | required by a bound introduced by this call
  |
  = note: for local types consider adding `#[derive(serde::Serialize)]` to your `NotSerializable<'_>` type
  = note: for types from other crates check whether the crate offers a `serde` feature flag
  = help: the following other types implement trait `Serialize`:
            &'a T
            &'a mut T
            ()
            (T,)
            (T0, T1)
            (T0, T1, T2)
            (T0, T1, T2, T3)
            (T0, T1, T2, T3, T4)
          and $N others
  = note: required for `(NotSerializable<'_>,)` to implement `Serialize`
  = note: this error originates in the macro `salsa::plumbing::setup_tracked_struct` which comes from the expansion of the attribute macro `salsa::tracked` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `NotSerializable<'_>: Deserialize<'_>` is not satisfied
 --> tests/compile-fail/invalid_serialize.rs:1:1
  |
1 | #[salsa::tracked(serialize)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Deserialize<'_>` is not implemented for `NotSerializable<'_>`
  |
  = note: for local types consider adding `#[derive(serde::Deserialize)]` to your `NotSerializable<'_>` type
  = note: for types from other crates check whether the crate offers a `serde` feature flag
  = help: the following other types implement trait `Deserialize<'de>`:
            &'a Path
            &'a [u8]
            &'a str
            ()
            (T,)
            (T0, T1)
            (T0, T1, T2)
            (T0, T1, T2, T3)
          and $N others
  = note: required for `(NotSerializable<'_>,)` to implement `Deserialize<'_>`
  = note: this error originates in the macro `salsa::plumbing::setup_tracked_struct` which comes from the expansion of the attribute macro `salsa::tracked` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `NotSerializable<'db>: Serialize` is not satisfied
  --> tests/compile-fail/invalid_serialize.rs:12:45
   |
12 | fn query(_db: &dyn salsa::Database, _input: NotSerializable<'_>) {}
   |                                             ^^^^^^^^^^^^^^^^^^^ the trait `Serialize` is not implemented for `NotSerializable<'db>`
   |
   = note: for local types consider adding `#[derive(serde::Serialize)]` to your `NotSerializable<'db>` type
   = note: for types from other crates check whether the crate offers a `serde` feature flag
   = help: the following other types implement trait `Serialize`:
             &'a T
             &'a mut T
             ()
             (T,)
             (T0, T1)
             (T0, T1, T2)
             (T0, T1, T2, T3)
             (T0, T1, T2, T3, T4)
           and $N others
note: required by a bound in `query_input_is_serializable`
  --> tests/compile-fail/invalid_serialize.rs:11:1
   |
11 | #[salsa::tracked(serialize)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `query_input_is_serializable`
   = note: this error originates in the macro `salsa::plumbing::setup_tracked_fn` which comes from the expansion of the attribute macro `salsa::tracked` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `NotSerializable<'db>: DeserializeOwned` is not satisfied
  --> tests/compile-fail/invalid_serialize.rs:12:45
   |
12 | fn query(_db: &dyn salsa::Database, _input: NotSerializable<'_>) {}
   |                                             ^^^^^^^^^^^^^^^^^^^ the trait `for<'de> Deserialize<'de>` is not implemented for `NotSerializable<'db>`
   |
   = help: the following other types implement trait `Deserialize<'de>`:
             &'a Path
             &'a [u8]
             &'a str
             ()
             (T,)
             (T0, T1)
             (T0, T1, T2)
             (T0, T1, T2, T3)
           and $N others
   = note: required for `NotSerializable<'db>` to implement `DeserializeOwned`
note: required by a bound in `query_input_is_serializable`
  --> tests/compile-fail/invalid_serialize.rs:11:1
   |
11 | #[salsa::tracked(serialize)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `query_input_is_serializable`
   = note: this error originates in the macro `salsa::plumbing::setup_tracked_fn` which comes from the expansion of the attribute macro `salsa::tracked` (in Nightly builds, run with -Z macro-backtrace for more info)
